core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,492 @@
+    // subchannel has already been created and attempted a connection, try the next one
+    } else {
+      addressIndex.increment();
+      if (addressIndex.isValid()) {
+        requestConnection();
+      }
+    }
+  }
+
+  /**
+  * Returns true if all other subchannels have already attempted and failed a connection so that
+  * the load balancer can report a state of TRANSIENT_FAILURE and stop serial connection attempts.
+*/
+  private boolean enterTransientFailure() {
+    for (ConnectivityState state : states.values()) {
+      if (state != TRANSIENT_FAILURE) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private Subchannel createNewSubchannel(SocketAddress addr) {
+    final Subchannel subchannel = helper.createSubchannel(
+        CreateSubchannelArgs.newBuilder()
+        .setAddresses(Lists.newArrayList(
+            new EquivalentAddressGroup(addr)))
+            .build());
+    subchannels.put(addr, subchannel);
+    states.put(subchannel, IDLE);
+    subchannel.start(new SubchannelStateListener() {
+      @Override
+      public void onSubchannelState(ConnectivityStateInfo stateInfo) {
+        processSubchannelState(subchannel, stateInfo);
+      }
+    });
+    return subchannel;
+  }
+
+  @VisibleForTesting
+  ConnectivityState getCurrentState() {
+    return this.currentState;
+  }
+
+  /**
+   * No-op picker which doesn't add any custom picking logic. It just passes already known result
+   * received in constructor.
+   */
+  private static final class Picker extends SubchannelPicker {
+    private final PickResult result;
+
+    Picker(PickResult result) {
+      this.result = checkNotNull(result, "result");
+    }
+
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      return result;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(Picker.class).add("result", result).toString();
+    }
+  }
+
+  /**
+   * Picker that requests connection during the first pick, and returns noResult.
+   */
+  private final class RequestConnectionPicker extends SubchannelPicker {
+    private final PickFirstLeafLoadBalancer pickFirstLeafLoadBalancer;
+    private final AtomicBoolean connectionRequested = new AtomicBoolean(false);
+
+    RequestConnectionPicker(PickFirstLeafLoadBalancer pickFirstLeafLoadBalancer) {
+      this.pickFirstLeafLoadBalancer =
+          checkNotNull(pickFirstLeafLoadBalancer, "pickFirstLeafLoadBalancer");
+    }
+
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      if (connectionRequested.compareAndSet(false, true)) {
+        helper.getSynchronizationContext().execute(new Runnable() {
+          @Override
+          public void run() {
+            pickFirstLeafLoadBalancer.updateBalancingState(CONNECTING,
+                new Picker(PickResult.withNoResult()));
+            pickFirstLeafLoadBalancer.requestConnection();
```

Comment:
Why do we need L389

Reply:
Resolved in meeting: wait for callback from subchannel